### PR TITLE
Share cached arrays

### PR DESCRIPTION
**Pull request by cquiroz**
First of a series of ITC performance PRs.

I was always suspicious at the amount of array cloning in ITC. My impression is that this was done for historical reasons out of fear of changing underlying data. A more careful examination confirmed it and for the most part we can make it immutable, avoiding quite a few allocations and array copies. 

Every spectrum built from a data file copied both arrays out of the `DatFile` cache, and `clone()` copied them again. Nothing ever mutates an array spectrum, so the mutating methods move from `ArraySpectrum` to `SampledSpectrum`, `DefaultArraySpectrum` becomes immutable and shares the cached arrays, and its `clone()` returns itself.

Please focus on the second commit for review, the first commit adds a test harness. It is similar in principle to the existing Baseline but it is focused on the json side and can gives as measurements to tell the performance improvements in cpu and memory while avoiding regressions.

| | main | this PR | change |
|---|---|---|---|
| allocation per run | 1936 MB | 825 MB | -57% |
| wall time per run | 3236 ms | 3097 ms | -4% |
| calspec IFU-2 case | 641 MB / 332 ms | 151 MB / 295 ms | -76% / -11% |

The third commit includes a bug fix:
rescaleX derived the redshifted sampling from (end - start) / n with n the number of samples rather than the n - 1 intervals, so redshifted results depended on the array length; it now uses sampling * (1 + z), which is correct but changed slightly the output thus a new baseline is included

### CODE REVIEW

File: bundle/edu.gemini.itc/src/main/java/edu/gemini/itc/base/DefaultArraySpectrum.java [hunk 1]
```diff
@@ -27,8 +29,18 @@ public final class DefaultArraySpectrum implements ArraySpectrum {
...
+
+    /** Wraps arrays that are owned elsewhere (typically the DatFile cache) without copying them. */
+    public static DefaultArraySpectrum sharing(final double[][] data) {
+        return new DefaultArraySpectrum(data, false);
+    }
+
+    private DefaultArraySpectrum(final double[][] data, final boolean copy) {
+        assert data.length == 2;
+        assert data[0].length == data[1].length;
+        _data = copy ? new double[][] { data[0].clone(), data[1].clone() } : data;
```
[the remark this answers is not in the document]
Comment (toddburnside) [reply bundle/edu.gemini.itc/src/main/java/edu/gemini/itc/base/DefaultArraySpectrum.java]: It seems this kind of eliminates the benefits you are trying to achieve? I guess the idea is a sort of agreement not to be changing the arrays?

**Review by toddburnside — APPROVED** [review]
Copilot has some comments, but otherwise LGTM.

File: bundle/edu.gemini.itc/src/main/java/edu/gemini/itc/base/DefaultArraySpectrum.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (cquiroz) [reply bundle/edu.gemini.itc/src/main/java/edu/gemini/itc/base/DefaultArraySpectrum.java]: Good point. `sharing` is now package-private, so only `Filter` can wrap cache arrays, and its doc says the arrays must not be modified afterwards. The public constructors keep copying their inputs.

Note the mutators no longer exist on `DefaultArraySpectrum`, so nothing can write through a shared instance; the only remaining way to corrupt the cache is to write to the raw `DatFile.arrays()` result

File: bundle/edu.gemini.itc.web/src/test/scala/edu/gemini/itc/web/golden/GoldenSnapshot.scala [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+          }
+        }
+    }
+    GoldenSnapshot(scalars.result(), strings.result(), series.result())
+  }
+
+  // ===== Comparison
+
+  private def close(a: Double, b: Double, relTol: Double): Boolean =
+    a == b || (a.isNaN && b.isNaN) || math.abs(a - b) <= relTol * math.max(math.abs(a), math.abs(b))
```
[the remark this answers is not in the document]
Comment (cquiroz) [reply bundle/edu.gemini.itc.web/src/test/scala/edu/gemini/itc/web/golden/GoldenSnapshot.scala]: It's a bit of an edge case but valid, adding a fix

File: bundle/edu.gemini.itc.web/src/test/scala/edu/gemini/itc/web/golden/ItcGoldenSpec.scala [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  val update: Boolean = setting("ITC_GOLDEN_UPDATE").exists(v => v == "1" || v.equalsIgnoreCase("true"))
+  val relTol: Double  = setting("ITC_GOLDEN_REL_TOL").map(_.toDouble).getOrElse(1e-12)
+  val filter: String  = setting("ITC_GOLDEN_FILTER").getOrElse("")
+
+  // Tests are forked from the project directory but the harness may run from the repo root.
+  val goldenFile: File = setting("ITC_GOLDEN_FILE").map(new File(_)).getOrElse(
+    List("src/test/resources/golden/itc-golden.json.gz", "bundle/edu.gemini.itc.web/src/test/resources/golden/itc-golden.json.gz")
+      .map(new File(_)).find(f => f.getParentFile.getParentFile.exists()).getOrElse(new File("itc-golden.json.gz")))
+
+  val cases: List[GoldenCase] = GoldenCases.all.filter(_.key.contains(filter))
```
[the remark this answers is not in the document]
Comment (cquiroz) [reply bundle/edu.gemini.itc.web/src/test/scala/edu/gemini/itc/web/golden/ItcGoldenSpec.scala]: Will add it
